base/compiler/typelattice.jl

```diff
@@ -56,6 +60,57 @@ end
 import Core: InterConditional
 const AnyConditional = Union{Conditional,InterConditional}
 
+# If an object field reference can be aliased to another reference of the same object field,
+# this lattice element records the object identity and wraps the field type. It then allows
+# certain built-in functions like `isa` and `===` to propagate a constraint imposed on the
+# field when `MustAlias` appears at a call-site.
+# This lattice element assumes the invariant that the field of wrapped slot object never
+# changes until the slot object is re-assigned. This means, the wrapped object should be
+# immutable since currently inference doesn't track any effects from memory writes.
+# `has_alias_field` takes the lift to check if a given lattice element is eligible to be
+# wrapped by `MustAlias`.
+# NOTE currently this lattice element is only used in abstractinterpret, not in optimization
+struct MustAlias
+    var::SlotNumber
+    vartyp::Any
+    fld::Const
+    fldtyp::Any
+    function MustAlias(
+                var::SlotNumber,
+                @nospecialize(vartyp),
+                fld::Const,
+                @nospecialize(fldtyp))
+        # avoid to be a slot wrapper of a slot wrapper (hard to maintain)
+        fldtyp = widenslotwrappers(fldtyp)
+        return new(var, vartyp, fld, fldtyp)
+    end
+end
+
+# This lattice element is very similar to `InterConditional`, but corresponds to `MustAlias`
+# struct InterMustAlias
+#     slot::Int
+#     fld::Const
+#     fldtyp::Any
+# end
+import Core: InterMustAlias
+const AnyMustAlias = Union{MustAlias,InterMustAlias}
+
+# XXX are these check really enough to hold invariants that `MustAlias` expects ?
+function has_alias_field(@nospecialize(t))
+    t = widenconst(t)
+    cnt = fieldcount_noerror(t)
+    (cnt === nothing || cnt == 0) && return false
+    return !ismutabletype(t)
```

Comment:
We need to change this check so that it syncs with [the `PartialStruct`'s construction check](https://github.com/JuliaLang/julia/blob/bc3ce48636b700d153d2c84614aaea8cb8843d43/base/compiler/abstractinterpretation.jl#L1431) since @vtjnash pointed out that some code assumes `PartialStruct.typ` to be concrete (but with some special casing for not-fully concrete tuples added).